torch/csrc/autograd/c2isl_function.cpp

```diff
@@ -0,0 +1,321 @@
+#include "torch/csrc/autograd/c2isl_function.h"
+#include "torch/csrc/autograd/functions/utils.h"
+
+#include "torch/csrc/utils/auto_gil.h"
+#include "torch/csrc/utils/auto_gpu.h"
+#include "torch/csrc/utils/object_ptr.h"
+
+// TODO: remove me when we have actual backwards
+#include "torch/csrc/autograd/functions/basic_ops.h"
+
+#include "THC/THC.h"
+
+extern THCState* state;
+
+namespace torch { namespace autograd {
+
+DLDataType toDLDataType(at::Type& ty) { // should be const
+  DLDataType t;
+  // Torch does not have any vectorized types.
+  t.lanes = 1;
+  switch (ty.scalarType()) {
+    case at::ScalarType::Byte:
+      t.code = DLDataTypeCode::kUInt;
+      t.bits = 8;
+      break;
+    case at::ScalarType::Char:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 8;
+      break;
+    case at::ScalarType::Double:
+      t.code = DLDataTypeCode::kFloat;
+      t.bits = 64;
+      break;
+    case at::ScalarType::Float:
+      t.code = DLDataTypeCode::kFloat;
+      t.bits = 32;
+      break;
+    case at::ScalarType::Int:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 32;
+      break;
+    case at::ScalarType::Long:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 64;
+      break;
+    case at::ScalarType::Short:
+      t.code = DLDataTypeCode::kInt;
+      t.bits = 16;
+      break;
+    case at::ScalarType::Half:
+      t.code = DLDataTypeCode::kFloat;
+      t.bits = 16;
+      break;
+    case at::ScalarType::NumOptions:
```

Comment:
ATen's ScalarType is an enum class, so if we add default we lose exhaustiveness checking. I'll add a comment to this effect.